# Expoloring Hacker News Posts
ここでは、数十万人が利用する人気の技術系サイトHacker Newsに投稿されている記事（通称："posts"）のデータセットの分析実習を行う。このデータセットは、およそ30万のpostsデータを含んでいる。今回は、そこから、読者からのコメントを受けていない記事を除き、更に、残ったpostsから無作為抽出を行い得られた2万個のpostsデータリストを扱う。

# 1. Reading the Data Set

In [1]:
from csv import reader
opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
hn = list(read_file)

#Display the first five rows of hn
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


# 2. Removing Headers from a List of Lists¶

In [2]:
headers = hn[0]
hn = hn[1:]

print(headers)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [3]:
print(hn[:5])

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


# 3. Extracting Ask HN and Show HN Posts

In [4]:
ask_posts = [] 
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(row)
    elif title.lower().startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)

In [5]:
print(len(ask_posts))

1744


In [6]:
print(len(show_posts))

1162


In [7]:
print(len(other_posts))

17194


# 4. Calculating the Average Number of Comments for Ask HN and Show HN Posts

In [8]:
#Ask HN
total_ask_comments = 0

for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments
    
avg_ask_comments = total_ask_comments / len(ask_posts)
print("avg_ask_comments:", round(avg_ask_comments, 1))

avg_ask_comments: 14.0


In [9]:
#Show HN
total_show_comments = 0

for row in show_posts:
    num_comments = int(row[4])
    total_show_comments += num_comments
    
avg_show_comments = total_show_comments / len(show_posts)
print("avg_show_comments:", round(avg_show_comments, 1))

avg_show_comments: 10.3


show postsとask postsのコメントの平均値はそれぞれ10.3と14.0であった。ask postsの方が平均値が大きい。show postsのコメントは記事に対する感想（反作用的コメント）、ask postsのコメントは質問への回答（積極的コメントもしくは利他的コメント）と考える。回答コメントの平均値の方が多いことは読者の積極性の高さを感じさせられる。

# 5. Calculate the Amount of Ask Posts Created per Hour 

In [10]:
import datetime as dt

result_list = []

for row in ask_posts:
    created_time = row[6]
    ask_comments = int(row[4])
    
    result_list.append([created_time, ask_comments])

In [16]:
counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    created_dt = row[0]
    created_dt = dt.datetime.strptime(created_dt, "%m/%d/%Y %H:%M")
    total_comments = row[1]
    
    created_hour = created_dt.strftime("%H")
    
    if created_hour not in counts_by_hour:
        counts_by_hour[created_hour] = 1
        comments_by_hour[created_hour] = row[1]
    else:
        counts_by_hour[created_hour] += 1
        comments_by_hour[created_hour] += row[1]

In [17]:
print(counts_by_hour)
print(comments_by_hour)

{'07': 34, '00': 55, '13': 85, '14': 107, '20': 80, '15': 116, '21': 109, '09': 45, '18': 109, '12': 73, '01': 60, '11': 58, '23': 68, '05': 46, '17': 100, '03': 54, '10': 59, '02': 58, '19': 110, '22': 71, '16': 108, '04': 47, '06': 44, '08': 48}
{'07': 267, '00': 447, '13': 1253, '14': 1416, '20': 1722, '15': 4477, '21': 1745, '09': 251, '18': 1439, '12': 687, '01': 683, '11': 641, '23': 543, '05': 464, '17': 1146, '03': 421, '10': 793, '02': 1381, '19': 1188, '22': 479, '16': 1814, '04': 337, '06': 397, '08': 492}


In [20]:
avg_by_hour = []

for hour in comments_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour] / counts_by_hour[hour]])
    
print(avg_by_hour)

[['07', 7.852941176470588], ['00', 8.127272727272727], ['13', 14.741176470588234], ['14', 13.233644859813085], ['20', 21.525], ['15', 38.5948275862069], ['21', 16.009174311926607], ['09', 5.5777777777777775], ['18', 13.20183486238532], ['12', 9.41095890410959], ['01', 11.383333333333333], ['11', 11.051724137931034], ['23', 7.985294117647059], ['05', 10.08695652173913], ['17', 11.46], ['03', 7.796296296296297], ['10', 13.440677966101696], ['02', 23.810344827586206], ['19', 10.8], ['22', 6.746478873239437], ['16', 16.796296296296298], ['04', 7.170212765957447], ['06', 9.022727272727273], ['08', 10.25]]


# 6. Sorting and Printing Values from a List of Lists

In [21]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
print(swap_avg_by_hour)

[[7.852941176470588, '07'], [8.127272727272727, '00'], [14.741176470588234, '13'], [13.233644859813085, '14'], [21.525, '20'], [38.5948275862069, '15'], [16.009174311926607, '21'], [5.5777777777777775, '09'], [13.20183486238532, '18'], [9.41095890410959, '12'], [11.383333333333333, '01'], [11.051724137931034, '11'], [7.985294117647059, '23'], [10.08695652173913, '05'], [11.46, '17'], [7.796296296296297, '03'], [13.440677966101696, '10'], [23.810344827586206, '02'], [10.8, '19'], [6.746478873239437, '22'], [16.796296296296298, '16'], [7.170212765957447, '04'], [9.022727272727273, '06'], [10.25, '08']]


In [22]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
print(sorted_swap[:5])

[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21']]


In [28]:
for row in sorted_swap:
    hour = dt.datetime.strptime(row[1], "%H")
    hour = hour.strftime("%H:%M")
    
    format_avg = "{}: {:.2f} average comments per post".format(hour, row[0])
    print(format_avg)

15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post
13:00: 14.74 average comments per post
10:00: 13.44 average comments per post
14:00: 13.23 average comments per post
18:00: 13.20 average comments per post
17:00: 11.46 average comments per post
01:00: 11.38 average comments per post
11:00: 11.05 average comments per post
19:00: 10.80 average comments per post
08:00: 10.25 average comments per post
05:00: 10.09 average comments per post
12:00: 9.41 average comments per post
06:00: 9.02 average comments per post
00:00: 8.13 average comments per post
23:00: 7.99 average comments per post
07:00: 7.85 average comments per post
03:00: 7.80 average comments per post
04:00: 7.17 average comments per post
22:00: 6.75 average comments per post
09:00: 5.58 average comments per post


各時間帯に送られる、ask post１つあたりの平均コメント数は15時台が最も多い。
また、16時台は4番目に多い。したがって、より多くのコメントをもらうためには15時少し前にpostsを投稿すべきである。